File: 프로그래머스/2/72411. 메뉴 리뉴얼/메뉴 리뉴얼.java
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Comparator;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Objects;
import java.util.Optional;

public class Solution {
	Map<Integer, Map<String, Integer>> combsByCourse = new HashMap<>();
	Map<Integer, Integer> maxCountByCourse = new HashMap<>();

	public void combination(char[] source, char[] comb, int start, int depth, int r) {
		if (depth == r) {
			int mergedCount = combsByCourse.computeIfAbsent(r, (key) -> new HashMap<>())
				.merge(new String(comb), 1, Integer::sum);
			maxCountByCourse.compute(r, (k, v) -> v == null ? mergedCount : Math.max(v, mergedCount));
			return;
		}

		for (int i = start; i <= source.length - (r - depth); i++) {
			comb[depth] = source[i];
			combination(source, comb, i + 1, depth + 1, r);
		}
	}

	public String[] solution(String[] orders, int[] courses) {

		for (String order : orders) {
			char[] orderDetail = order.toCharArray();
			Arrays.sort(orderDetail);

			for (int r : courses) {
				combination(orderDetail, new char[r], 0, 0, r);
			}
		}

		List<String> answer = new ArrayList<>();

		for (int course : courses) {
			Optional.ofNullable(maxCountByCourse.get(course))
				.filter(max -> max >= 2)
				.ifPresent(max -> {
					combsByCourse.get(course).entrySet().stream()
						.filter(entry -> Objects.equals(entry.getValue(), max))
						.forEach(entry -> answer.add(entry.getKey()));
				});
		}

		return answer.stream().sorted(Comparator.naturalOrder()).toArray(String[]::new);
	}
}

File: 프로그래머스/2/72411. 메뉴 리뉴얼/메뉴 리뉴얼.py
from collections import Counter
from typing import List
from itertools import combinations


def solution(orders: List[str], course: List[int]) -> List[str]:
    answer = []
    
    sorted_orders = [sorted(order) for order in orders]

    for course_item_count in course:
        course_combs: List[str] = []
        
        for order in sorted_orders:
            combs = combinations(order, course_item_count)
            course_combs += ["".join(comb) for comb in combs]

        counts = Counter(course_combs)
        
        if len(counts) == 0:
            continue
        
        max_count = max(counts.values())
        
        if max_count < 2:
            continue
        
        answer += [k for k, count in counts.items() if count == max_count]
        
    return sorted(answer)